### PR TITLE
Add tests for #20103

### LINKED ISSUE #20103: New type system: stackoverflow when defining method
I'm pretty sure this only showed up after the #18457 merge (maybe add to list in #19998):
```
    _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.2153 (2017-01-18 05:52 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit b8971ea4f* (0 days old master)
|__/                   |  x86_64-pc-linux-gnu

julia> immutable TT{X,Y} end

julia> f{X,Y}(::Type{TT{X,Y}},x::X,y::Y) = 1
f (generic function with 1 method)

julia> f{X}(::Type{TT{X,X}},x::X) = 100
ERROR: StackOverflowError:

julia> f(TT{Int,Int},1,2)
1

julia> f(TT{Int,Int},1) # it still works fine!
100

julia> methods(f)
# 2 methods for generic function "f":
f{X}(::Type{TT{X,X}}, x::X) in Main at REPL[3]:1
f{X,Y}(::Type{TT{X,Y}}, x::X, y::Y) in Main at REPL[2]:1
```

### PR DESCRIPTION
**Pull request by martinholters**
Although not everything is perfect with all the comments in #20103, most of them (and the OP) have been fixed eventually, so I'd say this closes #20103.

### CODE REVIEW

**Comment by martinholters** [comment]
Travis 64bit linux timed out (10 minutes without output), so probably a test hanging, but `subtype` did finish. Travis on MacOS failed in the `spawn` tests. Both failures look unrelated, so restarted after backing logs up at https://gist.github.com/martinholters/9383f043ea1acaae32050ee30e98d765.

File: test/subtype.jl [hunk 1]
```diff
@@ -1070,3 +1070,12 @@ end
 @testintersect(Tuple{A20992{R, D, d} where d where D, Int} where R,
                Tuple{C20992{S, n, T, D, d} where d where D where T where n where S, Any},
                Tuple{C20992, Int})
+
+# issue #20103, OP and comments
+immutable TT20103{X,Y} end
```
Comment (musm) [inline test/subtype.jl]: struct ?

File: test/subtype.jl [hunk 2]
```diff
@@ -1070,3 +1070,12 @@ end
...
+f20103{X,Y}(::Type{TT20103{X,Y}},x::X,y::Y) = 1
+f20103{X}(::Type{TT20103{X,X}},x::X) = 100
+@testintersect(Tuple{Int, Ref{Pair{K,V}}} where V where K,
```
Comment (JeffBezanson) [inline test/subtype.jl]: This test already appears on line 834.

File: test/subtype.jl (hunk 1 above)
Comment (martinholters) [inline test/subtype.jl]: yes

File: test/subtype.jl (hunk 2 above)
Comment (martinholters) [inline test/subtype.jl]: ah, thanks for pointing that out

**Comment by martinholters** [comment]
Comments addressed.

**Comment by martinholters** [comment]
AppVeyor failure unrelated (`GitError(Code:ERROR, Class:OS, Failed to receive response: The server returned an invalid or unrecognized response)`).

File: test/subtype.jl (hunk 1 above)
Comment (tkelman) [inline test/subtype.jl]: still says immutable

**Comment by martinholters** [comment]
Ok, this time I think I actually have addressed the comments.

**Pull request merged**